File: src/ts/graph/fuse.ts
/**
 * Graph lowerer: pure transform GraphDefinition → FusePlan (Spec 02).
 *
 * No wasm instantiation, no AOT compile. Output is the sole input to
 * multi-root / fused-tick codegen (Spec 03).
 *
 * ## Boundary rewrite policy (one policy, tested)
 *
 * **Keep real host port / param names** in each node's `expr` text.
 * Do **not** rewrite to GraphRunner's positional `x,y,z` aliases.
 *
 * Rationale: fused codegen maps ports by stable names across the module
 * (`nodeId.param` for flattened params; producer node ids for edge slots).
 * Multi-module GraphRunner rewrites only because single-node AOT is fixed to
 * positional `x/y/z` — that constraint does not apply to the fuse plan.
 *
 * ## Param naming
 *
 * Flattened fused entry params use `"${nodeId}.${param}"` (globally unique).
 * Order: topo order of reachable compute nodes, then each node's
 * `Object.keys(params)` insertion order.
 *
 * ## Port kinds
 *
 * Known `PortKind`s are accepted as plan metadata: number/boolean plus
 * matrix/complex/record/series/string/any (Spec 07 full-value). Unknown kind
 * strings throw via `normalizePortKind`. `inputKinds` is preserved on each
 * plan node (default `"number"`). Edge kind compatibility is checked lightly
 * (aligned with GraphRunner `kindsCompatible`) so the lowerer is a safe sole
 * gate for fuse paths.
 *
 * ## Outputs
 *
 * Empty/omitted `outputs` inherit `normalizeGraphDefinition`'s rule: every
 * node becomes an output (`id → id.out`). The dedicated empty-outputs error
 * only applies when normalize still yields no outputs (empty node list).
 *
 * ## v1 rejections
 *
 * - Reachable `type: "wasm"` nodes (opaque prebuilt modules)
 * - Missing edge for a required expr input port
 * - Undeclared input port, input+param name collision, non-finite param
 * - Duplicate host-facing input `name`s
 * - Output ref to a missing node / non-`.out` port
 * - Kind mismatch on an edge into a reachable expr node
 * - Cycles (via `topoSort`)
 */

import {
  normalizeGraphDefinition,
  nodeOutputKind,
  parseGraphRef,
  type GraphDefinition,
  type GraphNode,
  type GraphNodeId,
  type GraphValue,
  type PortKind,
} from "./schema";
import { topoSort, type GraphRefSource, type Topology } from "./topo";
import { kindsCompatible, normalizePortKind } from "./value_transfer";

// ── Stable error messages (testable substrings) ─────────────────────────────

export const FUSE_ERR = {
  wasmUnsupported: "Fuse v1 does not support wasm nodes",
  missingInputEdge: "missing edge for input",
  missingOutputNode: "output references missing node",
  emptyOutputs: "graph has no outputs",
  cycle: "Graph contains a cycle",
  undeclaredPort: "has undeclared input port",
  inputParamCollision: "as both input and param",
  nonFiniteParam: "must be a finite number",
  outputNotOut: "must reference '.out'",
  invalidRef: "Invalid graph reference",
  duplicateInputName: "duplicate graph input name",
  kindMismatch: "Kind mismatch on edge",
} as const;

export class FuseError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "FuseError";
  }
}

// ── FusePlan types ──────────────────────────────────────────────────────────

export type FusePlanInput = {
  /**
   * Host-facing input name (`node.name ?? node.id`).
   * Unique within `plan.inputs` (duplicate names rejected).
   */
  name: string;
  kind: PortKind;
  sourceNodeId: GraphNodeId;
};

export type FusePlanParam = {
  /** Globally unique flat name: `nodeId.param`. */
  name: string;
  kind: "number";
  nodeId: GraphNodeId;
  /** Local param key on the source node. */
  param: string;
  default: number;
};

export type FusePlanConst = {
  id: GraphNodeId;
  value: GraphValue;
  kind: PortKind;
};

/**
 * One reachable compute (`expr`) node in topo order.
 *
 * `expr` keeps real port names (see module policy).
 * `inputPorts[i]` is the producer **node id** feeding `inputs[i]` (edge binding).
 * `inputKinds[i]` is the consumer port kind (default number).
 * `paramNames` are local keys; flat fused names are in `FusePlan.params`.
 */
export type FusePlanNode = {
  id: GraphNodeId;
  expr: string;
  /** Declared input port names (order preserved from the graph node). */
  inputs: string[];
  /** Producer node id for each entry in `inputs` (parallel). */
  inputPorts: string[];
  /** Consumer port kinds parallel to `inputs` (default `"number"`). */
  inputKinds: PortKind[];
  /** Local param names on this node (order matches node.params keys). */
  paramNames: string[];
  outputKind: PortKind;
};

export type FusePlanOutput = {
  name: string;
  fromNodeId: GraphNodeId;
  kind: PortKind;
};

export type FusePlan = {
  inputs: FusePlanInput[];
  params: FusePlanParam[];
  /** Reachable const nodes only (feed producers for edge binding / codegen). */
  consts: FusePlanConst[];
  /** Topo order among reachable compute (expr) nodes only. */
  nodes: FusePlanNode[];
  outputs: FusePlanOutput[];
  /**
   * Documented boundary policy for consumers of this plan.
   * Always `"real_names"` in v1 (no x,y,z rewrite).
   */
  boundaryPolicy: "real_names";
};

// ── Lowerer ─────────────────────────────────────────────────────────────────

/**
 * Lower a graph definition to a deterministic fuse plan.
 * Throws {@link FuseError} on all lowerer-owned failures (including cycles).
 */
export function lowerGraphToFusePlan(def: GraphDefinition): FusePlan {
  const normalized = normalizeGraphDefinition(def);
  let topology: Topology;
  try {
    topology = topoSort(normalized.nodes, normalized.edges);
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e);
    // Match topo.ts prefix "Graph contains a cycle involving: …"
    if (msg.startsWith(FUSE_ERR.cycle)) throw new FuseError(msg);
    throw e instanceof Error ? e : new Error(msg);
  }

  const byId = new Map<GraphNodeId, GraphNode>();
  for (const node of normalized.nodes) byId.set(node.id, node);

  const outputEntries = Object.entries(normalized.outputs);
  // Only hit when normalize yields nothing (empty node list). Non-empty graphs
  // with omitted/empty outputs auto-fill every node as an output.
  if (outputEntries.length === 0) {
    throw new FuseError(FUSE_ERR.emptyOutputs);
  }

  // Resolve outputs first — missing refs are hard errors (no silent drop).
  const planOutputs: FusePlanOutput[] = [];
  const seedIds = new Set<GraphNodeId>();
  for (const [outName, ref] of outputEntries) {
    let parsed: { nodeId: string; port: string };
    try {
      parsed = parseGraphRef(ref);
    } catch (e) {
      const msg = e instanceof Error ? e.message : String(e);
      throw new FuseError(
        msg.startsWith(FUSE_ERR.invalidRef)
          ? `${msg} (output '${outName}')`
          : `${FUSE_ERR.invalidRef} '${ref}' (output '${outName}').`,
      );
    }
    if (parsed.port !== "out") {
      throw new FuseError(
        `Output '${outName}' ${FUSE_ERR.outputNotOut} (got '${ref}').`,
      );
    }
    const src = byId.get(parsed.nodeId);
    if (!src) {
      throw new FuseError(
        `${FUSE_ERR.missingOutputNode} '${parsed.nodeId}' (output '${outName}').`,
      );
    }
    seedIds.add(parsed.nodeId);
    planOutputs.push({
      name: outName,
      fromNodeId: parsed.nodeId,
      kind: nodeOutputKind(src),
    });
  }

  const reachable = markReachable(seedIds, topology.incoming, byId);

  // Reject wasm among reachable nodes (v1).
  for (const id of reachable) {
    const node = byId.get(id)!;
    if (node.type === "wasm") {
      throw new FuseError(
        `${FUSE_ERR.wasmUnsupported} (node '${id}'). Replace with an expr node or keep multi-module GraphRunner.`,
      );
    }
  }

  const inputs: FusePlanInput[] = [];
  const consts: FusePlanConst[] = [];
  const nodes: FusePlanNode[] = [];
  const params: FusePlanParam[] = [];
  const seenInputNames = new Set<string>();

  // Topo order among all nodes; emit only reachable of each kind.
  for (const node of topology.ordered) {
    if (!reachable.has(node.id)) continue;

    switch (node.type) {
      case "input": {
        const name = node.name ?? node.id;
        if (seenInputNames.has(name)) {
          const prior = inputs.find((i) => i.name === name)?.sourceNodeId ?? "?";
          throw new FuseError(
            `${FUSE_ERR.duplicateInputName} '${name}' (nodes '${prior}' and '${node.id}').`,
          );
        }
        seenInputNames.add(name);
        inputs.push({
          name,
          kind: nodeOutputKind(node),
          sourceNodeId: node.id,
        });
        break;
      }
      case "const": {
        consts.push({
          id: node.id,
          value: node.value,
          kind: nodeOutputKind(node),
        });
        break;
      }
      case "expr": {
        const declared = [...(node.inputs ?? [])];
        const incoming = topology.incoming.get(node.id) ?? new Map<string, GraphRefSource>();
        const inputPorts: string[] = [];
        const inputKinds: PortKind[] = declared.map((_, i) =>
          normalizePortKind(node.inputKinds?.[i] ?? "number"),
        );

        for (let i = 0; i < declared.length; i++) {
          const port = declared[i]!;
          const src = incoming.get(port);
          if (!src) {
            throw new FuseError(
              `Expr node '${node.id}' ${FUSE_ERR.missingInputEdge} '${port}'.`,
            );
          }
          inputPorts.push(src.nodeId);

          // Lightweight kind check (aligned with GraphRunner typeCheckEdges).
          const srcNode = byId.get(src.nodeId);
          if (srcNode) {
            const consumerKind = inputKinds[i]!;
            // Expr without explicit outputKind defaults to number (same as runner).
            const producerKind =
              srcNode.type === "expr" && !srcNode.outputKind
                ? ("number" as PortKind)
                : nodeOutputKind(srcNode);
            if (!kindsCompatible(producerKind, consumerKind)) {
              throw new FuseError(
                `${FUSE_ERR.kindMismatch} '${src.nodeId}.out' → '${node.id}.${port}': producer '${producerKind}' vs consumer '${consumerKind}'.`,
              );
            }
          }
        }

        // Undeclared ports with edges → error (same as GraphRunner policy).
        for (const port of incoming.keys()) {
          if (!declared.includes(port)) {
            throw new FuseError(
              `Expr node '${node.id}' ${FUSE_ERR.undeclaredPort} '${port}'.`,
            );
          }
        }

        const paramRecord = node.params ?? {};
        const paramNames = Object.keys(paramRecord);
        for (const p of paramNames) {
          if (declared.includes(p)) {
            throw new FuseError(
              `Expr node '${node.id}' uses '${p}' ${FUSE_ERR.inputParamCollision}.`,
            );
          }
          const defVal = paramRecord[p];
          if (typeof defVal !== "number" || !Number.isFinite(defVal)) {
            throw new FuseError(
              `Param '${node.id}.${p}' ${FUSE_ERR.nonFiniteParam}.`,
            );
          }
          params.push({
            name: `${node.id}.${p}`,
            kind: "number",
            nodeId: node.id,
            param: p,
            default: defVal,
          });
        }

        nodes.push({
          id: node.id,
          // Boundary policy: real names — store expression as authored.
          expr: node.expr,
          inputs: declared,
          inputPorts,
          inputKinds,
          paramNames,
          outputKind: nodeOutputKind(node),
        });
        break;
      }
      case "wasm":
        // Unreachable wasm is fine (dropped); reachable already rejected above.
        break;
    }
  }

  return {
    inputs,
    params,
    consts,
    nodes,
    outputs: planOutputs,
    boundaryPolicy: "real_names",
  };
}

/**
 * Backward reachability from graph outputs through edge producers (DFS stack).
 * Marks every node whose value can affect at least one declared output.
 * The reachable **set** is order-independent; only the set matters for emit.
 */
function markReachable(
  seeds: Set<GraphNodeId>,
  incoming: Map<GraphNodeId, Map<string, GraphRefSource>>,
  byId: Map<GraphNodeId, GraphNode>,
): Set<GraphNodeId> {
  const reachable = new Set<GraphNodeId>();
  const stack = [...seeds];
  while (stack.length > 0) {
    const id = stack.pop()!;
    if (reachable.has(id)) continue;
    if (!byId.has(id)) continue;
    reachable.add(id);
    const ports = incoming.get(id);
    if (!ports) continue;
    for (const src of ports.values()) {
      if (!reachable.has(src.nodeId)) stack.push(src.nodeId);
    }
  }
  return reachable;
}

File: src/graph/fuse.zig
//! Graph lowerer: pure transform NormalizedGraphDefinition → FusePlan (Spec 02 / 05).
//!
//! Mirrors `src/ts/graph/fuse.ts`. No wasm instantiation or AOT compile.
//! Output is the sole input to multi-root / fused-tick codegen (Spec 03).
//!
//! ## Boundary rewrite policy
//!
//! **Keep real host port / param names** in each node's `expr` text.
//! Do **not** rewrite to GraphRunner's positional `x,y,z` aliases.
//!
//! ## Param naming
//!
//! Flattened fused entry params use `"${nodeId}.${param}"` (globally unique).
//! Order: topo order of reachable compute nodes, then each node's params
//! insertion order.
//!
//! ## Port kinds
//!
//! Known `PortKind`s are accepted (number/boolean + matrix/complex/record/
//! series/string/any — Spec 07 full-value). Unknown kind strings fail at
//! schema parse via `normalizePortKind`.
//!
//! ## v1 rejections
//!
//! - Reachable `type: "wasm"` nodes (opaque prebuilt modules)
//! - Missing edge for a required expr input port
//! - Undeclared input port, input+param name collision, non-finite param
//! - Duplicate host-facing input `name`s
//! - Output ref to a missing node / non-`.out` port
//! - Kind mismatch on an edge into a reachable expr node
//! - Cycles (via `topoSort`)

const std = @import("std");
const schema = @import("schema.zig");
const topo = @import("topo.zig");
const Value = @import("../core/value.zig").Value;

// ── Stable error messages (testable substrings; match TS FUSE_ERR) ───────────

pub const ERR = struct {
    pub const wasm_unsupported = "Fuse v1 does not support wasm nodes";
    pub const missing_input_edge = "missing edge for input";
    pub const missing_output_node = "output references missing node";
    pub const empty_outputs = "graph has no outputs";
    pub const cycle = "Graph contains a cycle";
    pub const undeclared_port = "has undeclared input port";
    pub const input_param_collision = "as both input and param";
    pub const non_finite_param = "must be a finite number";
    pub const output_not_out = "must reference '.out'";
    pub const invalid_ref = "Invalid graph reference";
    pub const duplicate_input_name = "duplicate graph input name";
    pub const kind_mismatch = "Kind mismatch on edge";
};

pub const FuseError = error{
    WasmUnsupported,
    MissingInputEdge,
    MissingOutputNode,
    EmptyOutputs,
    Cycle,
    UndeclaredPort,
    InputParamCollision,
    NonFiniteParam,
    OutputNotOut,
    InvalidRef,
    DuplicateInputName,
    KindMismatch,
    OutOfMemory,
};

// ── FusePlan types ──────────────────────────────────────────────────────────

pub const FusePlanInput = struct {
    /// Host-facing input name (`node.name ?? node.id`). Unique within plan.inputs.
    name: []const u8,
    kind: schema.PortKind,
    source_node_id: []const u8,
};

pub const FusePlanParam = struct {
    /// Globally unique flat name: `nodeId.param`.
    name: []const u8,
    kind: schema.PortKind = .number,
    node_id: []const u8,
    /// Local param key on the source node.
    param: []const u8,
    default: f64,
};

pub const FusePlanConst = struct {
    id: []const u8,
    value: Value,
    kind: schema.PortKind,
};

/// One reachable compute (`expr`) node in topo order.
pub const FusePlanNode = struct {
    id: []const u8,
    /// Authored expression (real port names — boundary policy).
    expr: []const u8,
    /// Declared input port names (order preserved from the graph node).
    inputs: []const []const u8,
    /// Producer node id for each entry in `inputs` (parallel).
    input_ports: []const []const u8,
    /// Consumer port kinds parallel to `inputs` (default number).
    input_kinds: []const schema.PortKind,
    /// Local param names on this node (order matches node.params).
    param_names: []const []const u8,
    output_kind: schema.PortKind,
};

pub const FusePlanOutput = struct {
    name: []const u8,
    from_node_id: []const u8,
    kind: schema.PortKind,
};

pub const FusePlan = struct {
    inputs: []const FusePlanInput,
    params: []const FusePlanParam,
    /// Reachable const nodes only.
    consts: []const FusePlanConst,
    /// Topo order among reachable compute (expr) nodes only.
    nodes: []const FusePlanNode,
    outputs: []const FusePlanOutput,
    /// Always `"real_names"` in v1 (no x,y,z rewrite).
    boundary_policy: []const u8 = "real_names",
};

// ── Last error detail (CLI / tests) ─────────────────────────────────────────

var last_error_buf: [512]u8 = undefined;
var last_error_len: usize = 0;

pub fn lastError() []const u8 {
    return last_error_buf[0..last_error_len];
}

fn setError(comptime fmt: []const u8, args: anytype) void {
    const msg = std.fmt.bufPrint(&last_error_buf, fmt, args) catch blk: {
        const fallback = "fuse error (message too long)";
        @memcpy(last_error_buf[0..fallback.len], fallback);
        last_error_len = fallback.len;
        break :blk last_error_buf[0..fallback.len];
    };
    last_error_len = msg.len;
}

fn setErrorCopy(msg: []const u8) void {
    const n = @min(msg.len, last_error_buf.len);
    @memcpy(last_error_buf[0..n], msg[0..n]);
    last_error_len = n;
}

// ── Lowerer ─────────────────────────────────────────────────────────────────

/// Lower a normalized graph definition to a deterministic fuse plan.
/// All plan slices are allocated from `arena` (caller owns the arena).
/// Const Values in the plan are **not** retained (borrowed from `def`);
/// do not free const Values via the plan — use `schema.deinitConstValues` on def.
pub fn lowerGraphToFusePlan(
    arena: std.mem.Allocator,
    scratch: std.mem.Allocator,
    def: schema.NormalizedGraphDefinition,
) FuseError!FusePlan {
    last_error_len = 0;

    var topology = topo.topoSort(scratch, def.nodes, def.edges) catch |err| {
        if (err == error.Cycle) {
            const cycle_nodes = topo.takeLastCycleNodes();
            defer if (cycle_nodes.len > 0) scratch.free(cycle_nodes);
            if (cycle_nodes.len > 0) {
                setError("{s} involving: {s}", .{ ERR.cycle, cycle_nodes });
            } else {
                setErrorCopy(ERR.cycle);
            }
            return error.Cycle;
        }
        if (err == error.OutOfMemory) return error.OutOfMemory;
        setError("topo sort failed: {s}", .{@errorName(err)});
        return error.InvalidRef;
    };
    defer topology.deinit(scratch);

    var by_id = std.StringHashMap(schema.GraphNode).init(scratch);
    defer by_id.deinit();
    for (def.nodes) |node| {
        try by_id.put(node.id(), node);
    }

    if (def.outputs.len == 0) {
        setErrorCopy(ERR.empty_outputs);
        return error.EmptyOutputs;
    }

    // Resolve outputs first — missing refs are hard errors.
    var plan_outputs: std.ArrayListUnmanaged(FusePlanOutput) = .empty;
    errdefer plan_outputs.deinit(arena);

    var seed_ids = std.StringHashMap(void).init(scratch);
    defer seed_ids.deinit();

    for (def.outputs) |out| {
        const parsed = schema.parseGraphRef(out.ref) catch {
            setError("{s} '{s}' (output '{s}').", .{ ERR.invalid_ref, out.ref, out.name });
            return error.InvalidRef;
        };
        if (!std.mem.eql(u8, parsed.port, "out")) {
            setError("Output '{s}' {s} (got '{s}').", .{ out.name, ERR.output_not_out, out.ref });
            return error.OutputNotOut;
        }
        const src = by_id.get(parsed.node_id) orelse {
            setError("{s} '{s}' (output '{s}').", .{ ERR.missing_output_node, parsed.node_id, out.name });
            return error.MissingOutputNode;
        };
        try seed_ids.put(parsed.node_id, {});
        try plan_outputs.append(arena, .{
            .name = out.name,
            .from_node_id = parsed.node_id,
            .kind = src.outputKind(),
        });
    }

    var reachable = try markReachable(scratch, &seed_ids, &topology.incoming, &by_id);
    defer reachable.deinit();

    // Reject wasm among reachable nodes (v1).
    var rit = reachable.keyIterator();
    while (rit.next()) |id_ptr| {
        const node = by_id.get(id_ptr.*).?;
        if (node == .wasm) {
            setError("{s} (node '{s}'). Replace with an expr node or keep multi-module GraphRunner.", .{
                ERR.wasm_unsupported,
                id_ptr.*,
            });
            return error.WasmUnsupported;
        }
    }

    var inputs: std.ArrayListUnmanaged(FusePlanInput) = .empty;
    errdefer inputs.deinit(arena);
    var consts: std.ArrayListUnmanaged(FusePlanConst) = .empty;
    errdefer consts.deinit(arena);
    var nodes: std.ArrayListUnmanaged(FusePlanNode) = .empty;
    errdefer nodes.deinit(arena);
    var params: std.ArrayListUnmanaged(FusePlanParam) = .empty;
    errdefer params.deinit(arena);

    var seen_input_names = std.StringHashMap(void).init(scratch);
    defer seen_input_names.deinit();

    for (topology.ordered) |node| {
        if (!reachable.contains(node.id())) continue;

        switch (node) {
            .input => |n| {
                if (seen_input_names.contains(n.name)) {
                    var prior: []const u8 = "?";
                    for (inputs.items) |inp| {
                        if (std.mem.eql(u8, inp.name, n.name)) {
                            prior = inp.source_node_id;
                            break;
                        }
                    }
                    setError("{s} '{s}' (nodes '{s}' and '{s}').", .{
                        ERR.duplicate_input_name,
                        n.name,
                        prior,
                        n.id,
                    });
                    return error.DuplicateInputName;
                }
                try seen_input_names.put(n.name, {});
                try inputs.append(arena, .{
                    .name = n.name,
                    .kind = n.kind,
                    .source_node_id = n.id,
                });
            },
            .@"const" => |n| {
                try consts.append(arena, .{
                    .id = n.id,
                    .value = n.value,
                    .kind = n.kind,
                });
            },
            .expr => |n| {
                // Topo always inserts an empty incoming map per node.
                const incoming_ptr = topology.incoming.getPtr(n.id) orelse unreachable;

                var input_ports: std.ArrayListUnmanaged([]const u8) = .empty;
                errdefer input_ports.deinit(arena);

                for (n.inputs, 0..) |port, i| {
                    const src = incoming_ptr.get(port) orelse {
                        setError("Expr node '{s}' {s} '{s}'.", .{ n.id, ERR.missing_input_edge, port });
                        return error.MissingInputEdge;
                    };
                    try input_ports.append(arena, src.node_id);

                    // Lightweight kind check (aligned with GraphRunner / TS fuse).
                    if (by_id.get(src.node_id)) |src_node| {
                        const consumer_kind = n.input_kinds[i];
                        const producer_kind: schema.PortKind = switch (src_node) {
                            .expr => |en| en.output_kind, // schema defaults to .number
                            else => src_node.outputKind(),
                        };
                        if (!schema.kindsCompatible(producer_kind, consumer_kind)) {
                            setError("{s} '{s}.out' → '{s}.{s}': producer '{s}' vs consumer '{s}'.", .{
                                ERR.kind_mismatch,
                                src.node_id,
                                n.id,
                                port,
                                producer_kind.name(),
                                consumer_kind.name(),
                            });
                            return error.KindMismatch;
                        }
                    }
                }

                // Undeclared ports with edges → error.
                var port_it = incoming_ptr.keyIterator();
                while (port_it.next()) |port_ptr| {
                    const port = port_ptr.*;
                    var found = false;
                    for (n.inputs) |declared| {
                        if (std.mem.eql(u8, declared, port)) {
                            found = true;
                            break;
                        }
                    }
                    if (!found) {
                        setError("Expr node '{s}' {s} '{s}'.", .{ n.id, ERR.undeclared_port, port });
                        return error.UndeclaredPort;
                    }
                }

                var param_names: std.ArrayListUnmanaged([]const u8) = .empty;
                errdefer param_names.deinit(arena);

                for (n.params) |p| {
                    for (n.inputs) |declared| {
                        if (std.mem.eql(u8, declared, p.name)) {
                            setError("Expr node '{s}' uses '{s}' {s}.", .{
                                n.id,
                                p.name,
                                ERR.input_param_collision,
                            });
                            return error.InputParamCollision;
                        }
                    }
                    if (!std.math.isFinite(p.value)) {
                        setError("Param '{s}.{s}' {s}.", .{ n.id, p.name, ERR.non_finite_param });
                        return error.NonFiniteParam;
                    }
                    const flat = try std.fmt.allocPrint(arena, "{s}.{s}", .{ n.id, p.name });
                    try params.append(arena, .{
                        .name = flat,
                        .kind = .number,
                        .node_id = n.id,
                        .param = p.name,
                        .default = p.value,
                    });
                    try param_names.append(arena, p.name);
                }

                try nodes.append(arena, .{
                    .id = n.id,
                    .expr = n.expr,
                    .inputs = n.inputs,
                    .input_ports = try input_ports.toOwnedSlice(arena),
                    .input_kinds = n.input_kinds,
                    .param_names = try param_names.toOwnedSlice(arena),
                    .output_kind = n.output_kind,
                });
            },
            .wasm => {
                // Unreachable wasm is fine (dropped); reachable already rejected above.
            },
        }
    }

    return .{
        .inputs = try inputs.toOwnedSlice(arena),
        .params = try params.toOwnedSlice(arena),
        .consts = try consts.toOwnedSlice(arena),
        .nodes = try nodes.toOwnedSlice(arena),
        .outputs = try plan_outputs.toOwnedSlice(arena),
        .boundary_policy = "real_names",
    };
}

/// Backward reachability from graph outputs through edge producers.
fn markReachable(
    scratch: std.mem.Allocator,
    seeds: *const std.StringHashMap(void),
    incoming: *const std.StringHashMap(std.StringHashMap(topo.GraphRefSource)),
    by_id: *const std.StringHashMap(schema.GraphNode),
) error{OutOfMemory}!std.StringHashMap(void) {
    var reachable = std.StringHashMap(void).init(scratch);
    errdefer reachable.deinit();

    var stack: std.ArrayListUnmanaged([]const u8) = .empty;
    defer stack.deinit(scratch);

    var sit = seeds.keyIterator();
    while (sit.next()) |id| {
        try stack.append(scratch, id.*);
    }

    while (stack.items.len > 0) {
        const id = stack.pop().?;
        if (reachable.contains(id)) continue;
        if (!by_id.contains(id)) continue;
        try reachable.put(id, {});
        const ports = incoming.get(id) orelse continue;
        var pit = ports.valueIterator();
        while (pit.next()) |src| {
            if (!reachable.contains(src.node_id)) {
                try stack.append(scratch, src.node_id);
            }
        }
    }
    return reachable;
}

// Unit tests: tests/zig/graph/fuse_lowerer_test.zig